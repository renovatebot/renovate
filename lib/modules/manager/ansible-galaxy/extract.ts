import { logger } from '../../../logger';
import { newlineRegex, regEx } from '../../../util/regex';
import type { PackageDependency, PackageFileContent } from '../types';
import { extractCollections } from './collections';
import { extractCollectionsMetaDataFile } from './collections-metadata';
import { extractRoles } from './roles';

export function getSliceEndNumber(
  start: number,
  numberOfLines: number,
  ...blocks: number[]
): number {
  if (start < 0 || start > numberOfLines - 1) {
    return -1;
  }
  let nearestEnd = numberOfLines;
  for (const blocksKey of blocks) {
    if (start < blocksKey && blocksKey < nearestEnd) {
      nearestEnd = blocksKey;
    }
  }
  return nearestEnd;
}

export function extractPackageFile(
  content: string,
  packageFile: string
): PackageFileContent | null {
  logger.trace(`ansible-galaxy.extractPackageFile(${packageFile})`);
  const galaxyFileNameRegEx = regEx(/galaxy\.ya?ml$/);
  const deps: PackageDependency[] = [];
  const lines = content.split(newlineRegex);

  try {
    // if this is a galaxy.yml file we have to interpret the dependencies differently
    if (galaxyFileNameRegEx.exec(packageFile)) {
      const galaxyDeps = extractCollectionsMetaDataFile(lines);
      deps.push(...galaxyDeps);
    } else {
      // interpret requirements file
      // check if new or old format is used and save start lines for collection and roles.
      const positions = {
        collections: -1,
        roles: -1,
      };
      // find role and collection block
      lines.forEach((line, index) => {
        if (regEx(/^collections:/).exec(line)) {
          positions.collections = index;
        }
        if (regEx(/^roles:/).exec(line)) {
          positions.roles = index;
        }
      });
      if (positions.collections >= 0 || positions.roles >= 0) {
        // using new format
        const collectionLines = lines.slice(
          positions.collections,
          getSliceEndNumber(
            positions.collections,
            lines.length,
            positions.roles
          )
        );
        const collectionDeps = extractCollections(collectionLines);
        deps.push(...collectionDeps);

        const roleLines = lines.slice(
          positions.roles,
          getSliceEndNumber(
            positions.roles,
            lines.length,
            positions.collections
          )
        );
        const roleDeps = extractRoles(roleLines);
        deps.push(...roleDeps);
      } else {
        // use old format which only has only roles
        const galaxyDeps = extractRoles(lines);
        deps.push(...galaxyDeps);
      }
    }

    if (!deps.length) {
      return null;
    }
    return { deps };
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ err, packageFile }, 'Error extracting ansible-galaxy deps');
    return null;
  }
}
