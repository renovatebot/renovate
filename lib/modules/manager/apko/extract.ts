import { logger } from '../../../logger';
import { getSiblingFileName, readLocalFile } from '../../../util/fs';
import { parseSingleYaml } from '../../../util/yaml';
import { ApkDatasource } from '../../datasource/apk';
import { id as apkVersioning } from '../../versioning/apk';
import type { PackageDependency, PackageFileContent } from '../types';

// Simple schema for apko configuration
interface ApkoConfig {
  contents?: {
    repositories?: string[];
    packages?: string[];
  };
  environment?: Record<string, string>;
  cmd?: string;
  archs?: string[];
}

// Schema for apko lock file
interface ApkoLockFile {
  version: string;
  config: {
    name: string;
    checksum: string;
  };
  contents: {
    keyring: string[];
    build_repositories: string[];
    runtime_repositories: string[];
    repositories: {
      name: string;
      url: string;
      architecture: string;
    }[];
    packages: {
      name: string;
      url: string;
      version: string;
      architecture: string;
      checksum: string;
    }[];
  };
}

// apko allows arch alias, see https://github.com/chainguard-dev/apko/blob/5244a17ae460bb053b7d13302d48fd0046aeb387/pkg/apk/apk/arch.go
function translateArch(arch: string): string {
  switch (arch) {
    case 'i386':
    case '386':
      return 'x86';
    case 'amd64':
      return 'x86_64';
    case 'arm64':
      return 'aarch64';
    case 'arm/v6':
      return 'armhf';
    case 'arm/v7':
      return 'armv7';
    default:
      return arch;
  }
}

export async function extractPackageFile(
  content: string,
  packageFile: string,
): Promise<PackageFileContent | null> {
  logger.trace('apko.extractPackageFile()');

  try {
    const parsed = parseSingleYaml<ApkoConfig>(content);
    const deps: PackageDependency[] = [];

    // Check if archs is specified
    if (!parsed.archs) {
      throw new Error('archs must be specified');
    }

    // Extract packages from the contents.packages array
    if (parsed.contents?.packages) {
      // determine the registry URLs based on the archs
      const registryUrls = parsed.archs.map((arch) => {
        const translatedArch = translateArch(arch);
        return `${parsed.contents?.repositories?.[0]}/${translatedArch}`;
      });

      for (const pkg of parsed.contents.packages) {
        // Try to extract version from package.
        // format is name{@tag}{[<>~=]version} - from https://wiki.alpinelinux.org/wiki/Alpine_Package_Keeper#World
        const versionMatch = /^(.+)[=><~^][=]?(.+)$/.exec(pkg);

        if (versionMatch) {
          const [, depName, currentValue] = versionMatch;
          deps.push({
            datasource: ApkDatasource.id,
            depName,
            currentValue,
            versioning: apkVersioning,
            registryUrls,
          });
        } else {
          // Package without version - add as unversioned
          deps.push({
            datasource: ApkDatasource.id,
            depName: pkg,
            skipReason: 'not-a-version',
            registryUrls,
          });
        }
      }
    }

    if (!deps.length) {
      return null;
    }

    // Derive lockfile name from package file name
    // If package file is 'image.yaml', lockfile should be 'image.lock.json'
    const packageFileName =
      packageFile.split('/').pop() ?? 'apko.yaml'; /* v8 ignore next */
    const baseName = packageFileName.replace(/\.ya?ml$/, '');
    const lockFileName = getSiblingFileName(
      packageFile,
      `${baseName}.lock.json`,
    );
    const lockFileContent = await readLocalFile(lockFileName, 'utf8');

    let lockFiles: string[] | undefined;

    if (lockFileContent) {
      try {
        const lockFile = parseSingleYaml<ApkoLockFile>(lockFileContent);

        // Create a mapping of package names to locked versions
        const lockedVersions = new Map<string, string>();
        const lockedPackages = new Set<string>();

        // Process all packages from the lock file
        for (const pkg of lockFile.contents?.packages || []) {
          lockedVersions.set(pkg.name, pkg.version);
          lockedPackages.add(pkg.name);
        }

        // Add locked versions to dependencies that are in both files
        for (const dep of deps) {
          if (dep.depName && lockedVersions.has(dep.depName)) {
            dep.lockedVersion = lockedVersions.get(dep.depName);
            // Keep architecture-specific URLs for locked packages
          }
        }

        logger.debug(
          { packageFile, lockFileName, packageCount: deps.length },
          'Found apko.lock.json with locked versions',
        );
        lockFiles = [lockFileName];
      } catch (err) {
        logger.debug({ err, lockFileName }, 'Error parsing apko.lock.json');
        // Use architecture-specific repository URLs when lock file parsing fails
        const registryUrls =
          parsed.archs?.map((arch) => {
            const translatedArch = translateArch(arch);
            return `${parsed.contents?.repositories?.[0]}/${translatedArch}`;
          }) || parsed.contents?.repositories; /* v8 ignore next */

        for (const dep of deps) {
          dep.registryUrls = registryUrls;
        }
        lockFiles = ['not here'];
      }
    }

    return {
      deps,
      lockFiles,
    };
  } catch (err) {
    logger.debug({ err, packageFile }, 'Error parsing apko YAML configuration');
    return null;
  }
}
