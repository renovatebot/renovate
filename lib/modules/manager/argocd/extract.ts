import is from '@sindresorhus/is';
import { loadAll } from 'js-yaml';
import { logger } from '../../../logger';
import { coerceArray } from '../../../util/array';
import { trimTrailingSlash } from '../../../util/url';
import { DockerDatasource } from '../../datasource/docker';
import { GitTagsDatasource } from '../../datasource/git-tags';
import { HelmDatasource } from '../../datasource/helm';
import type {
  ExtractConfig,
  PackageDependency,
  PackageFileContent,
} from '../types';
import type {
  ApplicationDefinition,
  ApplicationSource,
  ApplicationSpec,
} from './types';
import { fileTestRegex } from './util';

export function extractPackageFile(
  content: string,
  packageFile: string,
  _config?: ExtractConfig
): PackageFileContent | null {
  // check for argo reference. API version for the kind attribute is used
  if (fileTestRegex.test(content) === false) {
    return null;
  }

  let definitions: ApplicationDefinition[];
  try {
    definitions = loadAll(content) as ApplicationDefinition[];
  } catch (err) {
    logger.debug({ err, packageFile }, 'Failed to parse ArgoCD definition.');
    return null;
  }

  const deps = definitions.filter(is.plainObject).flatMap(processAppSpec);

  return deps.length ? { deps } : null;
}

function processSource(source: ApplicationSource): PackageDependency | null {
  if (
    !source ||
    !is.nonEmptyString(source.repoURL) ||
    !is.nonEmptyString(source.targetRevision)
  ) {
    return null;
  }

  // a chart variable is defined this is helm declaration
  if (source.chart) {
    // assume OCI helm chart if repoURL doesn't contain explicit protocol
    if (
      source.repoURL.startsWith('oci://') ||
      !source.repoURL.includes('://')
    ) {
      let registryURL = source.repoURL.replace('oci://', '');
      registryURL = trimTrailingSlash(registryURL);

      return {
        depName: `${registryURL}/${source.chart}`,
        currentValue: source.targetRevision,
        datasource: DockerDatasource.id,
      };
    }

    return {
      depName: source.chart,
      registryUrls: [source.repoURL],
      currentValue: source.targetRevision,
      datasource: HelmDatasource.id,
    };
  }

  return {
    depName: source.repoURL,
    currentValue: source.targetRevision,
    datasource: GitTagsDatasource.id,
  };
}

function processAppSpec(
  definition: ApplicationDefinition
): PackageDependency[] {
  const spec: ApplicationSpec | null | undefined =
    definition.kind === 'Application'
      ? definition?.spec
      : definition?.spec?.template?.spec;

  if (is.nullOrUndefined(spec)) {
    return [];
  }

  const deps: (PackageDependency | null)[] = [];

  if (is.nonEmptyObject(spec.source)) {
    deps.push(processSource(spec.source));
  }

  for (const source of coerceArray(spec.sources)) {
    deps.push(processSource(source));
  }

  return deps.filter(is.truthy);
}
