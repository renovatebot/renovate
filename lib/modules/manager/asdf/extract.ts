import is from '@sindresorhus/is';
import { logger } from '../../../logger';
import { isSkipComment } from '../../../util/ignore';
import { regEx } from '../../../util/regex';
import type { PackageDependency, PackageFileContent } from '../types';
import { upgradeableTooling } from './upgradeable-tooling';

export function extractPackageFile(content: string): PackageFileContent | null {
  logger.trace(`asdf.extractPackageFile()`);

  const regex = regEx(
    /^(?<toolName>([\w_-]+)) (?<version>[^\s#]+)(?: [^\s#]+)* *(?: #(?<comment>.*))?$/gm
  );

  const deps: PackageDependency[] = [];

  for (const groups of [...content.matchAll(regex)]
    .map((m) => m.groups)
    .filter(is.truthy)) {
    const depName = groups.toolName.trim();
    const version = groups.version.trim();

    const toolConfig = upgradeableTooling[depName];
    const toolDefinition = toolConfig
      ? typeof toolConfig.config === 'function'
        ? toolConfig.config(version)
        : toolConfig.config
      : undefined;

    if (toolDefinition) {
      const dep: PackageDependency = {
        currentValue: version,
        depName,
        ...toolDefinition,
      };
      if (isSkipComment((groups.comment ?? '').trim())) {
        dep.skipReason = 'ignored';
      }

      deps.push(dep);
    } else {
      const dep: PackageDependency = {
        depName,
        skipReason: 'unsupported-datasource',
      };

      deps.push(dep);
    }
  }

  return deps.length ? { deps } : null;
}
