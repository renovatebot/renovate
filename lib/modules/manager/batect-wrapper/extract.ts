import { logger } from '../../../logger';
import { regEx } from '../../../util/regex';
import { GithubReleasesDatasource } from '../../datasource/github-releases';
import { id as semverVersioning } from '../../versioning/semver';
import type { PackageDependency, PackageFile } from '../types';

const VERSION_REGEX = regEx(/^\s+VERSION="(.*)"$/m);

export function extractPackageFile(fileContent: string): PackageFile | null {
  logger.trace('batect.extractPackageFile()');
  const match = VERSION_REGEX.exec(fileContent);

  if (match === null) {
    return null;
  }

  const dependency: PackageDependency = {
    depName: 'batect/batect',
    commitMessageTopic: 'Batect',
    currentValue: match[1],
    datasource: GithubReleasesDatasource.id,
    versioning: semverVersioning,
  };

  logger.trace(dependency, 'Found Batect wrapper version');

  return { deps: [dependency] };
}
