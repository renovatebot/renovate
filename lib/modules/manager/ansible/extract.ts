import { logger } from '../../../logger';
import { newlineRegex, regEx } from '../../../util/regex';
import * as dockerVersioning from '../../versioning/docker';
import { getDep } from '../dockerfile/extract';
import type {
  ExtractConfig,
  PackageDependency,
  PackageFileContent,
} from '../types';

export function extractPackageFile(
  content: string,
  packageFile: string,
  config: ExtractConfig
): PackageFileContent | null {
  logger.trace(`ansible.extractPackageFile(${packageFile})`);
  let deps: PackageDependency[] = [];
  const re = regEx(/^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/);
  for (const line of content.split(newlineRegex)) {
    const match = re.exec(line);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom, true, config.registryAliases);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.versioning = dockerVersioning.id;
      deps.push(dep);
    }
  }
  deps = deps.filter((dep) => !dep.currentValue?.includes('${'));
  if (!deps.length) {
    return null;
  }
  return { deps };
}
