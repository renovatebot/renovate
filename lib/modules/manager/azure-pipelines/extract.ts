import { load } from 'js-yaml';
import { GlobalConfig } from '../../../config/global';
import { logger } from '../../../logger';
import { coerceArray } from '../../../util/array';
import { regEx } from '../../../util/regex';
import { joinUrlParts } from '../../../util/url';
import { AzurePipelinesTasksDatasource } from '../../datasource/azure-pipelines-tasks';
import { GitTagsDatasource } from '../../datasource/git-tags';
import { getDep } from '../dockerfile/extract';
import type { PackageDependency, PackageFileContent } from '../types';
import type { AzurePipelines, Container, Repository } from './types';

const AzurePipelinesTaskRegex = regEx(/^(?<name>[^@]+)@(?<version>.*)$/);

export function extractRepository(
  repository: Repository
): PackageDependency | null {
  let repositoryUrl = null;

  if (repository.type === 'github') {
    repositoryUrl = `https://github.com/${repository.name}.git`;
  } else if (repository.type === 'git') {
    // "git" type indicates an AzureDevOps repository.
    // The repository URL is only deducible if we are running on AzureDevOps (so can use the endpoint)
    // and the name is of the form `Project/Repository`.
    // The name could just be the repository name, in which case AzureDevOps defaults to the
    // same project, which is not currently accessible here. It could be deduced later by exposing
    // the repository URL to managers.
    // https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-repositories-repository?view=azure-pipelines#types
    const { platform, endpoint } = GlobalConfig.get();
    if (platform === 'azure' && endpoint) {
      if (repository.name.includes('/')) {
        const [projectName, repoName] = repository.name.split('/');
        repositoryUrl = joinUrlParts(
          endpoint,
          encodeURIComponent(projectName),
          '_git',
          encodeURIComponent(repoName)
        );
      } else {
        logger.debug(
          'Renovate cannot update repositories that do not include the project name'
        );
      }
    }
  }

  if (repositoryUrl === null) {
    return null;
  }

  if (!repository.ref?.startsWith('refs/tags/')) {
    return null;
  }

  return {
    autoReplaceStringTemplate: 'refs/tags/{{newValue}}',
    currentValue: repository.ref.replace('refs/tags/', ''),
    datasource: GitTagsDatasource.id,
    depName: repository.name,
    depType: 'gitTags',
    packageName: repositoryUrl,
    replaceString: repository.ref,
  };
}

export function extractContainer(
  container: Container
): PackageDependency | null {
  if (!container.image) {
    return null;
  }

  const dep = getDep(container.image);
  logger.debug(
    {
      depName: dep.depName,
      currentValue: dep.currentValue,
      currentDigest: dep.currentDigest,
    },
    'Azure pipelines docker image'
  );
  dep.depType = 'docker';

  return dep;
}

export function extractAzurePipelinesTasks(
  task: string
): PackageDependency | null {
  const match = AzurePipelinesTaskRegex.exec(task);
  if (match?.groups) {
    return {
      depName: match.groups.name,
      currentValue: match.groups.version,
      datasource: AzurePipelinesTasksDatasource.id,
    };
  }
  return null;
}

export function parseAzurePipelines(
  content: string,
  packageFile: string
): AzurePipelines | null {
  let pkg: AzurePipelines | null = null;
  try {
    pkg = load(content, { json: true }) as AzurePipelines;
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ packageFile, err }, 'Error parsing azure-pipelines content');
    return null;
  }

  return pkg;
}

export function extractPackageFile(
  content: string,
  packageFile: string
): PackageFileContent | null {
  logger.trace(`azurePipelines.extractPackageFile(${packageFile})`);
  const deps: PackageDependency[] = [];

  const pkg = parseAzurePipelines(content, packageFile);
  if (!pkg) {
    return null;
  }

  for (const repository of coerceArray(pkg.resources?.repositories)) {
    const dep = extractRepository(repository);
    if (dep) {
      deps.push(dep);
    }
  }

  for (const container of coerceArray(pkg.resources?.containers)) {
    const dep = extractContainer(container);
    if (dep) {
      deps.push(dep);
    }
  }

  for (const { jobs } of coerceArray(pkg.stages)) {
    for (const { steps } of coerceArray(jobs)) {
      for (const step of coerceArray(steps)) {
        const task = extractAzurePipelinesTasks(step.task);
        if (task) {
          deps.push(task);
        }
      }
    }
  }

  for (const { steps } of coerceArray(pkg.jobs)) {
    for (const step of coerceArray(steps)) {
      const task = extractAzurePipelinesTasks(step.task);
      if (task) {
        deps.push(task);
      }
    }
  }

  for (const step of coerceArray(pkg.steps)) {
    const task = extractAzurePipelinesTasks(step.task);
    if (task) {
      deps.push(task);
    }
  }

  if (!deps.length) {
    return null;
  }
  return { deps };
}
