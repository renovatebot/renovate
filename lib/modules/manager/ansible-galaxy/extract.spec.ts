import { codeBlock } from 'common-tags';
import { Fixtures } from '../../../../test/fixtures';
import { getSliceEndNumber } from './extract';
import { extractPackageFile } from './';

const yamlFile1 = Fixtures.get('requirements01.yml');
const yamlFile2 = Fixtures.get('requirements02.yml');
const helmRequirements = Fixtures.get('helmRequirements.yml');
const collections1 = Fixtures.get('collections1.yml');
const collections2 = Fixtures.get('collections2.yml');
const galaxy = Fixtures.get('galaxy.yml');

describe('modules/manager/ansible-galaxy/extract', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty', () => {
      expect(extractPackageFile('nothing here', 'requirements.yml')).toBeNull();
    });

    it('extracts multiple dependencies from requirements.yml', () => {
      const res = extractPackageFile(yamlFile1, 'requirements.yml');
      expect(res?.deps).toMatchSnapshot();
      expect(res?.deps).toHaveLength(12);
    });

    it('extracts dependencies from a not beautified requirements file', () => {
      const res = extractPackageFile(yamlFile2, 'requirements.yml');
      expect(res?.deps).toMatchSnapshot();
      expect(res?.deps).toHaveLength(2);
    });

    it('extracts dependencies from requirements.yml with a space at the end of line', () => {
      const yamlFile = codeBlock`collections:
      - name: https://github.com/lowlydba/lowlydba.sqlserver.git
      type: git 
      version: 1.1.3`;
      const res = extractPackageFile(yamlFile, 'requirements.yml');
      expect(res?.deps).toHaveLength(1);
      expect(res?.deps[0].currentValue).toBe('1.1.3');
    });

    it('extracts git@ dependencies', () => {
      const yamlFile = codeBlock`collections:
      - name: community.docker
        source: git@github.com:ansible-collections/community.docker
        type: git
        version: 2.7.5`;
      const res = extractPackageFile(yamlFile, 'requirements.yml');
      expect(res?.deps).toHaveLength(1);
      expect(res?.deps[0].currentValue).toBe('2.7.5');
      expect(res?.deps[0].registryUrls).toBeUndefined();
      expect(res?.deps[0].packageName).toBe(
        'git@github.com:ansible-collections/community.docker'
      );
    });

    it('check if an empty file returns null', () => {
      const res = extractPackageFile('\n', 'requirements.yml');
      expect(res).toBeNull();
    });

    it('check if a requirements file of other systems returns null', () => {
      const res = extractPackageFile(helmRequirements, 'requirements.yml');
      expect(res).toBeNull();
    });

    it('check collection style requirements file', () => {
      const res = extractPackageFile(collections1, 'requirements.yml');
      expect(res?.deps).toMatchSnapshot();
      expect(res?.deps).toHaveLength(13);
      expect(res?.deps.filter((value) => value.skipReason)).toHaveLength(6);
    });

    it('check collection style requirements file in reverse order and missing empty line', () => {
      const res = extractPackageFile(collections2, 'requirements.yml');
      expect(res?.deps).toMatchSnapshot();
      expect(res?.deps).toHaveLength(4);
    });

    it('check galaxy definition file', () => {
      const res = extractPackageFile(galaxy, 'galaxy.yml');
      expect(res?.deps).toMatchSnapshot();
      expect(res?.deps).toHaveLength(2);
    });
  });

  describe('getSliceEndNumber()', () => {
    it('negative start number returns -1', () => {
      const res = getSliceEndNumber(-1, 10, 5);
      expect(res).toBe(-1);
    });

    it('a start number bigger then number of lines return -1', () => {
      const res = getSliceEndNumber(20, 10, 5);
      expect(res).toBe(-1);
    });

    it('choose first block', () => {
      const res = getSliceEndNumber(0, 10, 5);
      expect(res).toBe(5);
    });

    it('choose second block', () => {
      const res = getSliceEndNumber(5, 10, 5);
      expect(res).toBe(10);
    });
  });
});
