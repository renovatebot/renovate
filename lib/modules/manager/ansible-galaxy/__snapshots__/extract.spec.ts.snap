// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/ansible-galaxy/extract extractPackageFile() check collection style requirements file 1`] = `
[
  {
    "currentValue": "0.9.3",
    "datasource": "galaxy-collection",
    "depName": "geerlingguy.php_roles",
    "depType": "galaxy-collection",
    "registryUrls": [
      "https://galaxy.ansible.com",
    ],
  },
  {
    "currentValue": "1.2.2",
    "datasource": "galaxy-collection",
    "depName": "davidban77.gns3",
    "depType": "galaxy-collection",
    "registryUrls": [],
  },
  {
    "currentValue": "1.2.2",
    "datasource": "github-tags",
    "depName": "github.com/organization/repo_name",
    "depType": "galaxy-collection",
    "packageName": "https://github.com/organization/repo_name.git",
  },
  {
    "currentValue": "1.2.2",
    "datasource": "git-tags",
    "depName": "example.com/organization/repo_name",
    "depType": "galaxy-collection",
    "packageName": "https://example.com/organization/repo_name.git",
  },
  {
    "currentValue": "1.2.2",
    "datasource": "git-tags",
    "depName": "example.com/organization/repo_name",
    "depType": "galaxy-collection",
    "packageName": "https://example.com/organization/repo_name.git",
  },
  {
    "datasource": "galaxy-collection",
    "depName": "f5networks.f5_modules",
    "depType": "galaxy-collection",
    "registryUrls": [
      "https://cloud.redhat.com/api/automation-hub/",
    ],
    "skipReason": "unspecified-version",
  },
  {
    "currentValue": null,
    "datasource": "git-tags",
    "depName": "foo.bar/organization/repo_name",
    "depType": "galaxy-collection",
    "packageName": "https://foo.bar/organization/repo_name.git",
    "skipReason": "unspecified-version",
  },
  {
    "currentValue": "1.0.0",
    "depName": "fooBar",
    "depType": "galaxy-collection",
    "skipReason": "no-source-match",
  },
  {
    "depName": "foo.Bar",
    "depType": "galaxy-collection",
    "skipReason": "unsupported",
  },
  {
    "depName": "foo.Bar",
    "depType": "galaxy-collection",
    "skipReason": "local-dependency",
  },
  {
    "currentValue": "1.2.2",
    "datasource": "git-tags",
    "depName": "example.com/organization/repo_name",
    "depType": "galaxy-collection",
    "packageName": "https://example.com/organization/repo_name.git",
    "skipReason": "unsupported",
  },
  {
    "currentValue": "1.9.6",
    "datasource": "galaxy",
    "depName": "geerlingguy.java",
    "depType": "role",
    "packageName": "geerlingguy.java",
  },
  {
    "currentValue": "2.9.0",
    "datasource": "galaxy",
    "depName": "geerlingguy.docker",
    "depType": "role",
    "packageName": "geerlingguy.docker",
  },
]
`;

exports[`modules/manager/ansible-galaxy/extract extractPackageFile() check collection style requirements file in reverse order and missing empty line 1`] = `
[
  {
    "currentValue": "0.9.3",
    "datasource": "galaxy-collection",
    "depName": "geerlingguy.php_roles",
    "depType": "galaxy-collection",
    "registryUrls": [
      "https://galaxy.ansible.com",
    ],
  },
  {
    "currentValue": "1.2.2",
    "datasource": "galaxy-collection",
    "depName": "davidban77.gns3",
    "depType": "galaxy-collection",
  },
  {
    "currentValue": "1.9.6",
    "datasource": "galaxy",
    "depName": "geerlingguy.java",
    "depType": "role",
    "packageName": "geerlingguy.java",
  },
  {
    "currentValue": "2.9.0",
    "datasource": "galaxy",
    "depName": "geerlingguy.docker",
    "depType": "role",
    "packageName": "geerlingguy.docker",
  },
]
`;

exports[`modules/manager/ansible-galaxy/extract extractPackageFile() check galaxy definition file 1`] = `
[
  {
    "currentValue": "1.4.0",
    "datasource": "galaxy-collection",
    "depName": "ansible.windows",
    "depType": "galaxy-collection",
  },
  {
    "currentValue": ">=1.0.0,<2.0.0",
    "datasource": "galaxy-collection",
    "depName": "community.windows",
    "depType": "galaxy-collection",
  },
]
`;

exports[`modules/manager/ansible-galaxy/extract extractPackageFile() extracts dependencies from a not beautified requirements file 1`] = `
[
  {
    "currentValue": "0.1.0",
    "datasource": "galaxy",
    "depName": "yatesr.timezone",
    "depType": "role",
    "packageName": "yatesr.timezone",
  },
  {
    "currentValue": "0.1",
    "datasource": "git-tags",
    "depName": "mygroup/ansible-base",
    "depType": "role",
    "packageName": "git@gitlab.company.com:mygroup/ansible-base.git",
  },
]
`;

exports[`modules/manager/ansible-galaxy/extract extractPackageFile() extracts multiple dependencies from requirements.yml 1`] = `
[
  {
    "currentValue": "0.1.0",
    "datasource": "galaxy",
    "depName": "yatesr.timezone",
    "depType": "role",
    "packageName": "yatesr.timezone",
  },
  {
    "currentValue": "0.19.0",
    "datasource": "galaxy",
    "depName": "cloudalchemy.node-exporter",
    "depType": "role",
    "packageName": "cloudalchemy.node-exporter",
  },
  {
    "currentValue": "master",
    "datasource": "git-tags",
    "depName": "nginx_role",
    "depType": "role",
    "packageName": "https://github.com/bennojoy/nginx",
  },
  {
    "currentValue": "v1.4",
    "datasource": "git-tags",
    "depName": "willthames/git-ansible-galaxy",
    "depType": "role",
    "packageName": "http://bitbucket.org/willthames/git-ansible-galaxy",
  },
  {
    "currentValue": "0.1",
    "datasource": "git-tags",
    "depName": "mygroup/ansible-base",
    "depType": "role",
    "packageName": "git@gitlab.company.com:mygroup/ansible-base.git",
  },
  {
    "currentValue": "0.1",
    "datasource": "git-tags",
    "depName": "testGroup/testProject",
    "depType": "role",
    "packageName": "ssh://git@gitlab.company.com/testGroup/testProject.git",
  },
  {
    "currentValue": "0.1",
    "datasource": "git-tags",
    "depName": "testGroup/testProject2",
    "depType": "role",
    "packageName": "ssh://gitlab.company.com/testGroup/testProject2.git",
  },
  {
    "currentValue": "3.1",
    "datasource": "git-tags",
    "depName": "testGroup/testProject3",
    "depType": "role",
    "packageName": "ssh://git@gitlab.company.com:23/testGroup/testProject3.git",
  },
  {
    "currentValue": "0.14",
    "datasource": "git-tags",
    "depName": "mygroup/ansible-base",
    "depType": "role",
    "packageName": "git@gitlab.company.com:mygroup/ansible-base.git",
  },
  {
    "currentValue": "0.14",
    "datasource": "git-tags",
    "depName": "mygroup/ansible-base",
    "depType": "role",
    "packageName": "company.com:mygroup/ansible-base.git",
  },
  {
    "currentValue": "47.11",
    "datasource": "git-tags",
    "depName": "org/repo",
    "depType": "role",
    "packageName": "git://github.com/org/repo.git",
  },
  {
    "currentValue": "47.11",
    "datasource": "git-tags",
    "depName": "org/re.po",
    "depType": "role",
    "packageName": "git://github.com/org/re.po.git",
  },
]
`;
